cdmtaskservice/kafka_notifications.py

```diff
@@ -72,15 +84,31 @@ async def update_job_state(self, job_id: str, state: JobState, time: datetime.da
             "state": _not_falsy(state, "state").value,
             "time": _not_falsy(time, "time").isoformat()
         }).encode("utf-8"))
+        self._futures.add(future)  # ensure future isn't garbage collected
         # kafka client appears to enter an infinite loop if kafka is down
-        # may need to make this configurable, but 10s is a pretty long time to wait for a 
-        # tiny message to send. Don't worry about it for now
+        # will eventually send the messages if kafka comes back up
         # https://github.com/aio-libs/aiokafka/issues/1101
-        await asyncio.wait_for(future, 10)  # throw exception, if any. This is a pain to test
+        def cb(future: Future):
+            self._futures.discard(future)
+            try:
+                future.result()  # trigger exception and don't call callback if one occurs
+            except Exception:
+                # no idea how to test this, or what would trigger it
+                logging.getLogger(__name__).exception(
+                    f"Failed to send state update to Kafka for job {job_id}, state {state.value}"
+                )
+                return
+            if callback:
+                task = asyncio.create_task(callback)
+                self._tasks.add(task)  # prevent garbage collection
+                task.add_done_callback(self._tasks.discard)
+        future.add_done_callback(cb)
 
     async def close(self):
         """
         Close the notifier.
         """
         await self._prod.stop()
         self._closed = True
+        self._futures = None  # allow garbage collection
+        self._tasks = None  # allow garbage collection
```

Comment:
We don't need to call something like `self._tasks.discard(task)` or `task.cancel()` to clean up existing/stale tasks, right?

Reply:
We could if we wanted to be super careful, but this is only called on process shutdown so I wasn't worried about it. It just means the future callback / task might fire between this method being called and the GC running or the process exiting, so NBD.